### ISSUE TITLE
[no-unnecessary-type-assertion] false positive on uninitialized variables

### ISSUE THREAD

**Issue by mysticatea** [typescript-eslint/typescript-eslint#453]
**Repro**

```JSON
{
  "rules": {
    "@typescript-eslint/no-unnecessary-type-assertion": "error"
  }
}
```

```TS
function test() {
    let resolve: () => void
    new Promise(resolve0 => {
        resolve = resolve0
    })
    return resolve!
}
```

**Expected Result**

No errors because it's a syntax error if I removed the `!`.

**Actual Result**

```
6:12  error  This assertion is unnecessary since it does not change the type of the expression  @typescript-eslint/no-unnecessary-type-assertion
```

**Additional Info**

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `1.7.0` |
| `@typescript-eslint/parser`        | `1.7.0` |
| `TypeScript`                       | `3.4.3` |
| `ESLint`                           | `5.16.0` |
| `node`                             | `10.15.3` |
| `npm`                              | `6.9.0` |

**Comment by bradzacher** [comment]
I'm not sure that we can do anything about this.
The syntax error is because TS has decided the variable is uninitialised.
However "uninitialised" is not a type.
At the return statement, the variable has the type `() => void`, which is why there is an error - you're non-nulling a non-null type.
The operator is just telling ts to ignore the uninitialised state.

I haven't tested it, but I'd assume that the following code works:
```ts
function test() {
    let resolve: () => void | null = null
    new Promise(resolve0 => {
        resolve = resolve0
    })
    return resolve!
}
```

**Comment by mysticatea** [comment]
That rule has autofix and the autofix broke the code. **ESLint must not fix code if it changes semantics.** Also, to tell the variable was actually initialized is a intentional feature of `!`. It's definitely not "unnecessary assertion" and the rule should not warn it.
